alphazero/_game.py

```diff
@@ -383,3 +384,339 @@ def n_players():
+                check_1_win.append(sum(temp_1[k : k + 4]))
+
+        return (4 in check_0_win, 4 in check_1_win)
+
+    @staticmethod
+    def is_game_over(board) -> bool:
+        """
+        Check if the board state represents a game that is over. This should
+        just return a boolean True/False.
+        """
+        result = Connect4.result(board)
+        if result is None:
+            return False
+        # implicit else
+        return True
+
+    @staticmethod
+    def result(board) -> Union[None, int]:
+        """
+        Get the result of the game. This should be 1 for a win for the player
+        whose perspective we're looking for, 0 for a draw, -1 for a loss, or
+        None for undetermined, when the game is not over yet.
+        """
+        status = Connect4.get_game_status(board)
+
+        zero_win = status[0]
+        one_win = status[1]
+
+        if zero_win:
+            result = 1
+        elif one_win:
+            result = -1
+        elif torch.any(board.eq(-1)):
+            return None
+        else:
+            result = 0
+
+        return result
+
+    @property
+    def width(self) -> int:
+        """
+        Get the width of the game board. For tic-tac-toe, for instance, this
+        should return 3; for chess, 8.
+        """
+        return self._board.shape[1]
+
+    @property
+    def height(self) -> int:
+        """
+        Get the height of the game board. For tic-tac-toe, this should return
+        3; for chess, 8.
+        """
+        return self._board.shape[0]
+
+    @property
+    def board_state(self) -> torch.Tensor:
+        """
+        Get the current board state as a torch.Tensor.
+        """
+        return self._board
+
+    @classmethod
+    def from_json(cls, **params):
+        """
+        Create an instance of class `cls` starting from a JSON style dict
+        `params`. This dict should contain all the width, height, board state
+        parameters that may be needed to create an instance of this game class.
+        """
+        return cls(**params)
+
+    def reset(self) -> None:
+        """
+        Reset the game to the initial board position (as given by
+        `get_initial_board()`). Reset move lists, move counters, anything else.
+        """
+        self._board = self.get_initial_board()
+        self._move_count = 0
+        assert self.is_valid()
+
+    def is_valid(self) -> bool:
+        """
+        Check if the current board position is valid.
+        """
+        valid = True
+
+        zeroes = (self._board == 0).sum().item()
+        ones = (self._board == 1).sum().item()
+
+        if zeroes + ones != self._move_count:
+            valid = False
+
+        if abs(zeroes - ones) > 1:
+            valid = False
+
+        return valid
+
+    @staticmethod
+    def mirror_h(board):
+        """
+        Mirror the board position horizontally. Should be implemented per
+        subclass.
+        """
+        return board.flip(1)
+
```

Comment:
maybe implement mirror_v and raise a more specific error ('invalid for connect4' or whatever)?